### PR TITLE
Add Payment Controller and Payment Route

### LINKED ISSUE #79: BE: Payments Controller
Changes include the following:

- Model for payments needs to be completed first.
- New routes for payments.
   - `/payments/` <- Gets all payments for the signed in user.
   - `/payments/:paymentId` <- Gets a payment.
   - `/payments/:paymentId/pay` <- Make a payment for the service.
   - `/payments/:paymentId/cancel` <- Cancels a payment intent (**only the pet sitter "merchant" can do this)** 
- Controllers for all these routes. Think about middleware too and how it can help us, particularly with fetching profile, determining if the requester is of the right account type, access authroization, **input validation**, etc.
- Make sure to keep things DRY (Don't Repeat Yourself) in this controller.

### PR DESCRIPTION
**Pull request by RetiFier**
### What this PR does (required):
- Get Request for all payment and single payment
- Put Request for make payment and cancel payment

### Any information needed to test this feature (required):
- Need to add data manually 
- http://localhost:3001/payments for Create Payment
- http://localhost:3001/payments for Get All Payments
- http://localhost:3001/payments/:id Get single Payments
- http://localhost:3001/payments/:id/pay Make Payment with Put Method
- http://localhost:3001/payments/:Id/cancel Cancel Payment with Put Method
- Merchant user can only cancel

### CODE REVIEW

File: server/controllers/payment.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    });
+  }
+});
+
+exports.cancelPayment =  asyncHandler(async (req, res, next) => {
+  const userId = req.user.id;
+  const {id} = req.params;
+  const merchantPayment = await  Payment.findOne({id: id, sitterId: userId});
+  if (!merchantPayment) {
+    res.status(203);
```
Comment (sundayezeilo) [inline server/controllers/payment.js]: I think this should be a 401 ("Unauthorized") status

File: server/controllers/payment.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+          msg: "Payment has been Canceled"
+        }
+      });
+    }
+  }
+})
+
+exports.getAllPayments = asyncHandler(async (req, res, next) => {
+const payment = await Payment.find({userId: req.user.id});
+if(payment){
```
Comment (sundayezeilo) [inline server/controllers/payment.js]: There's no need for this check as this query will always return an array, which is always truthy, even when it's an empty array.

**Review by sundayezeilo — CHANGES_REQUESTED** [review]
This is cool. Just a few comment to have a look

File: server/controllers/payment.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+const asyncHandler = require("express-async-handler");
+const Payment = require("../models/Payment");
+
+
+
+exports.getPayment = asyncHandler(async (req, res, next) => {
+  const {id} = req.params;
```
Comment (moffatethan) [inline server/controllers/payment.js]:
```suggestion
  const { id } = req.params;
```

File: server/controllers/payment.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  if(payment){
+    res.status(200).json({
+      success: {   
+        payment: payment 
+      }
+    });
+  } else {
+    res.status(404);
+    throw new Error("Pls enter valid payment id");
+  }
```
Comment (moffatethan) [inline server/controllers/payment.js]: Could we add some whitespace above this conditional. Also, Could we do a different approach.

```js
if (!payment) {
   . . .
}

return res.status(200).json({  . . . });
```

Comment (moffatethan) [reply server/controllers/payment.js]: Ditto throughout the controller

File: server/controllers/payment.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+});
+
+exports.cancelPayment =  asyncHandler(async (req, res, next) => {
+  const userId = req.user.id;
+  const {id} = req.params;
+  const merchantPayment = await  Payment.findOne({id: id, sitterId: userId});
+  if (!merchantPayment) {
+    res.status(203);
+    throw new Error("Only Merchant Can Cancle");
```
Comment (moffatethan) [inline server/controllers/payment.js]:
```suggestion
    throw new Error("Only merchant can cancel");
```

File: server/controllers/payment.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  }  else { 
+    merchantPayment.set(({paid: false}))
```
Comment (moffatethan) [inline server/controllers/payment.js]: Can we do a quick schema change and add a cancel boolean to the payment and set that to true so we can filter that out on our queries. 🤔

File: server/routes/payment.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+const express = require('express');
+const { getAllPayments, getPayment, makePayment, cancelPayment } = require('../controllers/payment');
+const router = express.Router();
+const { protect } = require('../middleware/auth');
+
+router.route("/").get(protect, getAllPayments);
```
Comment (moffatethan) [inline server/routes/payment.js]: We need a post to create a payment too 👍🏻

**Review by moffatethan — CHANGES_REQUESTED** [review]
Just some changes to do, overall decent work.

**Review by moffatethan — CHANGES_REQUESTED** [review]
Address merge conflicts please

**Comment by RetiFier** [comment]
> Address merge conflicts please

@moffatethan  Resolved Conflicts

**Review by moffatethan — CHANGES_REQUESTED** [review]
Please address merge conflicts

**Review by moffatethan — APPROVED** [review]

**Pull request merged**